### PR TITLE
cmd/k8s-operator,k8s-operator,go.{mod,sum}: make individual proxy images/image pull policies configurable

### PR DESCRIPTION
**Pull request by irbekrm**
See tailscale/tailscale#11675 for context - this PR attempts to make it easier to configure what images are used for the operator proxies and make it possible to configure image pull policies.

Adds a couple new fields to `ProxyClass` CRD to make it possilble to configure images/image pull policies for tailscale container and tailscale init container:

```
apiVersion: tailscale.com/v1alpha1
kind: ProxyClass
...
spec:
...
  statefulSet:
    pod:
      tailscaleContainer:
        image: "ghcr.io/tailscale/tailscale:v1.64.0"
        imagePullPolicy: IfNotPresent
      tailscaleInitContainer:
        image: "ghcr.io/tailscale/tailscale:v1.64.0"
        imagePullPolicy: IfNotPresent
```

To try this out:
- build a new operator image
- apply CRDs from this PR `kubectl apply -f ./cmd/k8s-operator/deploy/crds`
- deploy the operator skipping the CRD and with the new image `helm upgrade --install operator tailscale-dev/tailscale-operator -n tailscale --set operatorConfig.image.repo=<image> --set installCRDs=false --set operatorConfig.image.tag=<tag>...`
- apply a `ProxyClass` with images/image pull policies configured, i.e `kubectl apply -f ./cmd/k8s-operator/examples/proxyclass.yaml`
- apply the `ProxyClass` to any operator's managed proxy  https://tailscale.com/kb/1236/kubernetes-operator#cluster-resource-customization-using-proxyclass-custom-resource
- observe that `Image`/`ImagePullPolicy` fields of the proxy `Pod` have the expected values

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -22,6 +22,7 @@ require (
 	github.com/dave/patsy v0.0.0-20210517141501-957256f50cba
 	github.com/dblohm7/wingoes v0.0.0-20240119213807-a09d6be7affa
 	github.com/digitalocean/go-smbios v0.0.0-20180907143718-390a4f403a8e
+	github.com/distribution/reference v0.6.0
```
Comment (irbekrm) [inline go.mod:25]: We also have https://github.com/tailscale/tailscale/blob/v1.64.2/go.mod#L184 which is a deprectated version of the same library pulled in by some dependency of mkctr it seems https://github.com/tailscale/mkctr/blob/main/go.mod#L13. We should probably update it there if possible

File: k8s-operator/apis/v1alpha1/types_proxyclass.go [hunk 2]
```diff
@@ -149,15 +149,29 @@ type Metrics struct {
...
+	// available at ghcr.io/tailscale/tailscale. Image name provided here
+	// will override any proxy image values specified via the Kubernetes
+	// operator's Helm chart values or PROXY_IMAGE env var to the operator
+	// deployment.
+	// https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#image
+	// +optional
+	Image string `json:"image,omitempty"`
+	// Image pull policy. One of Always, Never, IfNotPresent. Defaults to Always.
+	// https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#image
+	// +kubebuilder:validation:Enum=Always;Never;IfNotPresent
```
Comment (irbekrm) [inline k8s-operator/apis/v1alpha1/types_proxyclass.go:172]: > kubebuilder:validation:Enum=Always;Never;IfNotPresent

This ensures that any `ProxyClass` with an ImagePullPolicy value other than Always/Never/IfNotPresent gets rejected at apply time

File: k8s-operator/apis/v1alpha1/types_proxyclass.go [hunk 3]
```diff
@@ -149,15 +149,29 @@ type Metrics struct {
...
-	SecurityContext *corev1.SecurityContext `json:"securityContext,omitempty"`
+	Env []Env `json:"env,omitempty"`
+	// Container image name. By default images are pulled from
+	// docker.io/tailscale/tailscale, but the official images are also
+	// available at ghcr.io/tailscale/tailscale. Image name provided here
+	// will override any proxy image values specified via the Kubernetes
+	// operator's Helm chart values or PROXY_IMAGE env var to the operator
+	// deployment.
+	// https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#image
+	// +optional
```
Comment (irbekrm) [inline k8s-operator/apis/v1alpha1/types_proxyclass.go:168]: This is just me shifting these fields around to order them alphabetically (just to make it easier to read the code)

File: k8s-operator/apis/v1alpha1/types_proxyclass.go [hunk 4]
```diff
@@ -166,15 +180,15 @@ type Container struct {
...
+	// Container security context.
+	// Security context specified here will override the security context by the operator.
+	// By default the operator:
+	// - sets 'privileged: true' for the init container
+	// - set NET_ADMIN capability for tailscale container for proxies that
+	// are created for Services or Connector.
+	// https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#security-context
 	// +optional
-	Env []Env `json:"env,omitempty"`
+	SecurityContext *corev1.SecurityContext `json:"securityContext,omitempty"`
```
Comment (irbekrm) [inline k8s-operator/apis/v1alpha1/types_proxyclass.go:191]: This is just me shifting those fields around to order them alphabetically (just to make it easier to read the code)

**Comment by oxtoacart** [comment]
Given the below proxy class

```
apiVersion: tailscale.com/v1alpha1
kind: ProxyClass
metadata:
  name: prod
spec:
  statefulSet:
    annotations:
      platform-component: infra 
    pod:
      tailscaleContainer:
        env:
          - name: FOO
            value: BAZ
      labels:
        team: eng
      nodeSelector:
        beta.kubernetes.io/os: "linux"
      imagePullSecrets:
      - name: "foo"
      tailscaleContainer:
        image: "ghcr.io/tailscale/tailscale:v1.64.0"
        imagePullPolicy: IfNotPresent
      tailscaleInitContainer:
        image: "ghcr.io/tailscale/tailscale:v1.64.0"
        imagePullPolicy: IfNotPresent
```

In my testing, this worked for an ingress defined as follows:

```
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: kuardfoobar 
  labels:
    tailscale.com/proxy-class: "prod"
spec:
  tls:
  - hosts:
    - "dnstest"
  rules:
  - http:
      paths:
      - backend:
          service:
            name: kuardfoobar
            port:
              number: 80
        pathType: Prefix 
        path: /
  ingressClassName: tailscale
```

But, it did not work for a service defined as follows:

```
apiVersion: v1
kind: Service
metadata:
  name: kuardfoobar
  labels:
    tailscale.com/proxy-class: "prod"
spec:
  ports:
  - port: 80
    targetPort: 8080
    protocol: TCP
  selector:
    app: kuardfoobar
```

```
➜  oss git:(irbekrm/docim) ✗ kubectl get pod -n tailscale ts-kuardfoobar-k2pz9-0 -o yaml | grep image
    image: ghcr.io/tailscale/tailscale:v1.64.0
    imagePullPolicy: IfNotPresent
  imagePullSecrets:
    image: ghcr.io/tailscale/tailscale:v1.64.0
    imageID: ghcr.io/tailscale/tailscale@sha256:0e7dd9e2772c32958e9d92c06ab9ec696c74075a8d558b93eca5bc25dcc9dd19

➜  oss git:(irbekrm/docim) ✗ kubectl get pod -n tailscale ts-theservice-7vmgw-0 -o yaml | grep image
    image: us-central1-docker.pkg.dev/tailscale-sandbox/percy-images/proxy-dev:v0.0.11
    imagePullPolicy: Always
    image: us-central1-docker.pkg.dev/tailscale-sandbox/percy-images/proxy-dev:v0.0.11
    imagePullPolicy: IfNotPresent
    image: us-central1-docker.pkg.dev/tailscale-sandbox/percy-images/proxy-dev:v0.0.11
    imageID: us-central1-docker.pkg.dev/tailscale-sandbox/percy-images/proxy-dev@sha256:b9c479af8e0e1ec8edadcc75ea004afed12fd3354d9c7ca62e463c997bc94845
    image: us-central1-docker.pkg.dev/tailscale-sandbox/percy-images/proxy-dev:v0.0.11
    imageID: us-central1-docker.pkg.dev/tailscale-sandbox/percy-images/proxy-dev@sha256:b9c479af8e0e1ec8edadcc75ea004afed12fd3354d9c7ca62e463c997bc94845
```

**Comment by irbekrm** [comment]
> But, it did not work for a service defined as follows:

The `Service` resource definition in the example does not seem to have a [tailscale load balancer class](https://tailscale.com/kb/1236/kubernetes-operator#exposing-a-cluster-workload-via-a-tailscale-load-balancer-service) or [an annotation for the Tailscale Kubernetes operator](https://tailscale.com/kb/1236/kubernetes-operator#exposing-a-cluster-workload-by-annotating-an-existing-service) so it does not seem that the operator would have been creating any proxies for it.

> kubectl get pod -n tailscale ts-theservice-7vmgw-0

Judging from the `Pod` name, this is likely for a different `Service`?

File: k8s-operator/apis/v1alpha1/types_proxyclass.go [hunk 5]
```diff
@@ -151,15 +151,29 @@ type Metrics struct {
...
+	// the future.
 	// +optional
-	SecurityContext *corev1.SecurityContext `json:"securityContext,omitempty"`
+	Env []Env `json:"env,omitempty"`
+	// Container image name. By default images are pulled from
+	// docker.io/tailscale/tailscale, but the official images are also
+	// available at ghcr.io/tailscale/tailscale. Image name provided here
+	// will override any proxy image values specified via the Kubernetes
+	// operator's Helm chart values or PROXY_IMAGE env var to the operator
+	// deployment.
```
Comment (tomhjp) [inline k8s-operator/apis/v1alpha1/types_proxyclass.go:168]: nit: Minor rewording for clarity

```suggestion
	// available at ghcr.io/tailscale/tailscale. Specifying image name here
	// will override any proxy image values specified via the Kubernetes
	// operator's Helm chart values or PROXY_IMAGE env var in the operator
	// stateful set.
```

File: cmd/k8s-operator/deploy/examples/proxyclass.yaml [hunk 6]
```diff
@@ -15,3 +15,9 @@ spec:
         kubernetes.io/os: "linux"
       imagePullSecrets:
       - name: "foo"
+      tailscaleContainer:
+        image: "ghcr.io/tailscale/tailscale:v1.64.0"
+        imagePullPolicy: IfNotPresent
+      tailscaleInitContainer:
+        image: "ghcr.io/tailscale/tailscale:v1.64.0"
```
Comment (tomhjp) [inline cmd/k8s-operator/deploy/examples/proxyclass.yaml:22]: nit: Maybe either update to the latest tag (1.66.4) as of today, or use stable?

File: cmd/k8s-operator/proxyclass.go [hunk 7]
```diff
@@ -111,6 +112,21 @@ func (a *ProxyClassReconciler) validate(pc *tsapi.ProxyClass) (violations field.
...
 				}
+				if tc.Image != "" {
+					// Same validation as used by kubelet https://github.com/kubernetes/kubernetes/blob/release-1.30/pkg/kubelet/images/image_manager.go#L212
+					if _, err := dockerref.ParseNormalizedNamed(tc.Image); err != nil {
+						violations = append(violations, field.TypeInvalid(field.NewPath("spec", "statefulSet", "pod", "tailscaleContainer", "image"), tc.Image, err.Error()))
+
+					}
+				}
+			}
+			if tc := pod.TailscaleContainer; tc != nil {
```
Comment (tomhjp) [inline cmd/k8s-operator/proxyclass.go:123]: I think this is meant to select the init container?

File: cmd/k8s-operator/proxyclass.go [hunk 8]
```diff
@@ -111,6 +112,21 @@ func (a *ProxyClassReconciler) validate(pc *tsapi.ProxyClass) (violations field.
 						a.recorder.Event(pc, corev1.EventTypeWarning, reasonCustomTSEnvVar, fmt.Sprintf(messageCustomTSEnvVar, string(e.Name), "tailscale"))
 					}
 				}
+				if tc.Image != "" {
+					// Same validation as used by kubelet https://github.com/kubernetes/kubernetes/blob/release-1.30/pkg/kubelet/images/image_manager.go#L212
+					if _, err := dockerref.ParseNormalizedNamed(tc.Image); err != nil {
+						violations = append(violations, field.TypeInvalid(field.NewPath("spec", "statefulSet", "pod", "tailscaleContainer", "image"), tc.Image, err.Error()))
+
```
Comment (tomhjp) [inline cmd/k8s-operator/proxyclass.go:119]:
```suggestion
```

File: cmd/k8s-operator/proxyclass.go [hunk 9]
```diff
@@ -111,6 +112,21 @@ func (a *ProxyClassReconciler) validate(pc *tsapi.ProxyClass) (violations field.
 						a.recorder.Event(pc, corev1.EventTypeWarning, reasonCustomTSEnvVar, fmt.Sprintf(messageCustomTSEnvVar, string(e.Name), "tailscale"))
 					}
 				}
+				if tc.Image != "" {
+					// Same validation as used by kubelet https://github.com/kubernetes/kubernetes/blob/release-1.30/pkg/kubelet/images/image_manager.go#L212
```
Comment (tomhjp) [inline cmd/k8s-operator/proxyclass.go:116]: Nice!

**Review by tomhjp — COMMENTED** [review]
LGTM! I think there's just one minor bug in the ProxyClass validate function

File: k8s-operator/apis/v1alpha1/types_proxyclass.go (hunk 5 above)
Comment (irbekrm) [reply k8s-operator/apis/v1alpha1/types_proxyclass.go:168]: >  stateful set.

The operator itself is deployed via a `Deployment` https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/deploy/chart/templates/deployment.yaml, it's just the proxies that are `StatefulSet`s

File: cmd/k8s-operator/deploy/examples/proxyclass.yaml (hunk 6 above)
Comment (irbekrm) [reply cmd/k8s-operator/deploy/examples/proxyclass.yaml:22]: I'll put `1.64` in 👍🏼 This is just an example, but I generally find static image tags to be not a great practice

File: k8s-operator/apis/v1alpha1/types_proxyclass.go (hunk 5 above)
Comment (tomhjp) [reply k8s-operator/apis/v1alpha1/types_proxyclass.go:168]: Doh, of course

File: cmd/k8s-operator/proxyclass.go (hunk 7 above)
Comment (irbekrm) [reply cmd/k8s-operator/proxyclass.go]: Thanks, good catch! Fixed + added a test case for this

**Review by tomhjp — APPROVED** [review]
LGTM

**Pull request merged**